sklearn/base.py

```diff
@@ -133,7 +142,121 @@ def _pprint(params, offset=0, printer=repr):
     return lines
 
 
-class BaseEstimator:
+class _PropsRequest:
+    def get_props_request(self):
+        """Get requested data properties.
+
+        Returns
+        -------
+        props : dict of list of strings, or dict of dict of {str: str}, or None
+            The key to the top level dict is the method for which the prop is
+            used. Under each key, there is a dict list of required properties,
+            or a dict of mapping of the form
+            ``{method_parameter: provided_parameter}``, or None.
+        """
+        try:
+            return self._props_request
+        except AttributeError:
+            return {}
+
+    def set_props_request(self, props):
+        """Set required data properties.
+
+        Parameters
+        ----------
+        props : dict of list of strings, or dict of dict of {str: str}, or None
+            The key to the top level dict is the method for which the prop is
+            used. Under each key, there is a dict list of required properties,
+            or a dict of mapping of the form
+            ``{method_parameter: provided_parameter}``.
+            ``None`` empties the required props.
+
+        Returns
+        -------
+        self : estimator
+            returns self.
+        """
+        if props is None:
+            try:
+                del self._props_request
+            except AttributeError:
+                pass
+            return self
+
+        if not isinstance(props, dict):
+            raise ValueError("`props` should be a dictionary")
+
+        for method, value in props.items():
+            if isinstance(value, str):
+                props[method] = [value]
+
+        if not hasattr(self, '_props_request'):
+            self._props_request = {}
+
+        for method, m_props in props.items():
+            if method not in self._props_request:
+                self._props_request[method] = []
```

Comment:
I think we'll need to validate these methods here... I.e. consumers will need to have a set of valid methods, whether we have code to sniff it or it's explicitly set by each estimator.